src/lib.rs

```diff
@@ -156,50 +156,67 @@ pub fn uring_builder() -> io_uring::Builder {
     io_uring::IoUring::builder()
 }
 
-/// Builder API to allow starting the runtime and creating the io_uring driver with non-default
-/// parameters.
+/// Builder API that can create and start the `io_uring` runtime with non-default parameters,
+/// while abstracting away the underlying io_uring crate.
 // #[derive(Clone, Default)]
 pub struct Builder {
-    entries: u32,
+    sq_entries: u32,
+    cq_entries: u32,
     urb: io_uring::Builder,
 }
 
-/// Return a Builder to allow setting parameters before calling the start method.
-/// Returns a Builder with our default values, all of which can be replaced with the methods below.
+/// Constructs a [`Builder`] with default settings.
+/// Use this to alter submission and completion queue parameters, and to `start` the io_uring runtime.
 ///
-/// Refer to Builder::start for an example.
+/// Refer to [`Builder::start`] for an example.
 pub fn builder() -> Builder {
     Builder {
-        entries: 256,
+        sq_entries: 256,
+        cq_entries: 512,
         urb: io_uring::IoUring::builder(),
     }
 }
 
 impl Builder {
-    /// Set number of submission queue entries in uring.
+    /// Sets the number of submission queue entries in uring. The default is 256.
     ///
-    /// The kernel will ensure it uses a power of two and will round this up if necessary.
-    /// The kernel requires the number of completion queue entries to be larger than
-    /// the submission queue entries so generally will double the sq entries count.
+    /// The kernel requires the number of submission queue entries to be a power of two,
+    /// and that it be less than the number of completion queue entries.
+    /// This function will adjust the `cq_entries` value to be at least 2 times `sq_entries`
+    pub fn sq_entries(&mut self, sqe: u32) -> &mut Self {
+        self.sq_entries = sqe.next_power_of_two();
+        if self.cq_entries / 2 < self.sq_entries {
+            self.cq_entries = self.sq_entries * 2;
+        }
+        self.urb.setup_cqsize(self.cq_entries);
```

Comment:
I also can't get behind using this value to set a value in the default urb here. But maybe this is fixing something that was broken and I'm mistaken?

Reply:
It wasn't broken per-se, but the crate::Builder is just a thin wrapper over the io_uring::Builder, and io_uring::Builder.build() takes 1 parameter, the SQ Entries parameter.  It doesn't offer any other interface to set SQ entries size, but it does offer a method to set the CQ Entries size.  

Also, the Builders are used on start() in a couple different ways, so I wanted to be sure that the cq entries setting was saved before  io_uring::Builder.build() is called. 

What I think needs to happen is we need to refactor the io_uring::Builder API a bit, because there is no foolproof way to use it. 


https://github.com/rrichardson/tokio-uring/blob/47342a872fb90ea82236f9a506bdafc67036b424/src/runtime/driver/mod.rs#L40-L41